### DISCUSSION TITLE
What is a Shape Ptr?

### DISCUSSION THREAD

**Discussion by DoeringChristian** [mitsuba-renderer/mitsuba3#882] [Q&A] [answered]
Hi,

When casting a variable of type `ShapePtr` to an `UInt32` using `dr.reinterpret_array_v`, we seem to get an index back.
What is the meaning of that index and into what array would that be indexing?
Could we use it to index into an array of our own, for example to store data related to the shape?

For example in the following reproducer,
```python
import mitsuba as mi
import drjit as dr

mi.set_variant("cuda_ad_rgb")


if __name__ == "__main__":
    scene: mi.Scene = mi.load_dict(mi.cornell_box())

    print(f"{dr.reinterpret_array_v(mi.UInt, scene.shapes_dr())=}")

    ray = mi.Ray3f(mi.Point3f(0, 0, 0), mi.Vector3f(0, -1, 0))
    si: mi.SurfaceInteraction3f = scene.ray_intersect(ray)
    print(f"{dr.reinterpret_array_v(mi.UInt, si.shape)=}")
```
we got `[5, 6, 7, 1, 3, 8, 2, 4]` for the values of the `shapes_dr` array.
We also got 7 for the value of `si.shape` which, when looking up the 3rd shape seems to be the same index.
This seems to be consistent throughout runs.

Thanks for your Help

**Comment by njroussel** [comment, marked as answer]
Hi @DoeringChristian 

>  What is the meaning of that index and into what array would that be indexing?

Internally, Dr.Jit holds a registry of pointers ([documentation in source code)](https://github.com/mitsuba-renderer/drjit-core/blob/master/include/drjit-core/jit.h#L468-L497). Whenever a plugin is instantiated in Mitsuba it is registered with Dr.Jit and hence it's corresponding "vectorized pointer" is the index in that registry.

> Could we use it to index into an array of our own, for example to store data related to the shape?

Yes. Currently, it is guaranteed that a plugin will keep the same register index for however long it lives. Of course, if some plugin is deleted and a new one created, the new plugin might re-use the old one's index.

----

In short, whenever a plugin is instantiated it seems itself assigned an index for the entirety of its lifetime. I hope  this makes sense.

**Reply by DoeringChristian** [reply]
Thanks for the fast reply.
